### PR TITLE
Feature method not allowed

### PR DESCRIPTION
**Pull request by mrfishball**

### CODE REVIEW

**Comment by ericmeyer** [comment]
TLDR:
Your commit messages and history need work. I recommend following the conventions outlined [here](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html)

Long version:
A few comments regarding your commits:
1. You have a mix of tense. I recommend sticking to the imperative. e.g. "Fix Bug" instead of "Fixed Bug"
2. I would avoid merge commits in feature branches. You should be able to rebase your work on top of any new commits on master.
   * Note: merge commits can be unavoidable if/when you are working on a shared branch that you are not able to force push to.
3. This PR has failing commits. Again, you should make sure the build is passing before pushing up a commit.
4. Some of your commit messages are a bit too terse
   * e.g. "removed the use of forced unwrapping"
   * Removing from where? What changed?
   * Also, for this specific commit, wasn't there an actually feature change/addition? The message implies it was a refactoring.

**Comment by ericmeyer** [comment]
In general, I recommend taking a very disciplined approach to commits. You should feel free to come up with your own list, but mine looks something like this:
1. Make isolated change
2. Runs tests. Make sure they all run and pass.
3. Commit
4. Review git history

In a sense, this is kind of like the red-green-refactor cycle, but with commits.
 * Red: Uncommitted changes
 * Green: Tests all pass. All changes committed
 * Refactor: Rebase and clean up Git history for your branch

File: Sources/SwerverCore/Action/HeadAction.swift [hunk 1]
```diff
@@ -1,7 +1,7 @@
 public class HeadAction: HttpAction {
 
     private let responseBuilder = ResponseBuilder()
-
+
```
Comment (ericmeyer) [inline Sources/SwerverCore/Action/HeadAction.swift:4]: Remove trailing whitespace. In fact, most editors have a feature to automatically trim whitespace on save.

File: Sources/SwerverCore/Action/OptionsAction.swift [hunk 2]
```diff
@@ -21,9 +21,12 @@ public class OptionsAction: HttpAction {
     
     private func getAllowedMethods() -> String {
         var allowedMethodList: [String] = []
-        for (method, _) in routes.fetchRoute(url: route) {
-            allowedMethodList.append(method.rawValue)
+        if let actions = routes.fetchRoute(url: route)?.actions {
```
Comment (ericmeyer) [inline Sources/SwerverCore/Action/OptionsAction.swift]: Determining the available actions for a given route feels like the responsibility of the `Routes` class.

I would consider refactoring to introduce something like this:
```swift
class Routes {
    func allowedMethods(url: URL) -> [HttpMethod] {
    }
}
```

File: Sources/SwerverCore/Router/Route.swift [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        }
+        return listOfMethods
+    }
+    
+    public func notAllowedAction() -> HttpAction {
+        return NotAllowedAction(methods: getListOfMethods())
+    }
+}
+
+extension Route: Equatable {
```
Comment (ericmeyer) [inline Sources/SwerverCore/Router/Route.swift:26]: This is a good use of `extension`

File: Sources/SwerverCore/Router/Router.swift [hunk 4]
```diff
@@ -16,24 +16,33 @@ public class Router {
...
-                guard let action = targetRoute[targetMethod] else {
-                    return notFoundAction.execute()
+                guard let action = targetRoute.actions[targetMethod] else {
+                    return targetRoute.notAllowedAction().execute()
                 }
                 return action.execute()
+            } else {
+                return notFoundAction.execute()
             }
+        } else {
```
Comment (ericmeyer) [inline Sources/SwerverCore/Router/Router.swift:28]: :+1:

File: Sources/SwerverCore/Router/Routes.swift [hunk 5]
```diff
@@ -2,19 +2,19 @@ import Foundation
 
 public class Routes {
 
-    private var allRoutes: [URL:[RequestMethod:HttpAction]] = [:]
+    private var allRoutes: [URL:Route] = [:]
```
Comment (ericmeyer) [inline Sources/SwerverCore/Router/Routes.swift:5]: I really like the introduction of a `Route` abstraction.

File: Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+        describe("A Not Allowed action") {
+            let listOfAllowedMethods = ["GET", "POST"]
+            
+            let notAllowedAction = NotAllowedAction(methods: listOfAllowedMethods)
+            
+            it("can dispatch to response builder to build a response for a successful OPTIONS request") {
+                
+                let okResponse = notAllowedAction.execute()
+                expect(okResponse.statusCode).to(equal(StatusCode.not_allowed.rawValue))
+                expect(okResponse.headers[.contentType]).to(equal(ContentType.text.rawValue))
```
Comment (ericmeyer) [inline Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift]: I would not expose the underlying data structure for the headers. Instead, I would do `response.get(header: .allow)`

File: Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+class NotAllowedActionSpec: QuickSpec {
+    
+    override func spec() {
+        describe("A Not Allowed action") {
+            let listOfAllowedMethods = ["GET", "POST"]
+            
+            let notAllowedAction = NotAllowedAction(methods: listOfAllowedMethods)
+            
+            it("can dispatch to response builder to build a response for a successful OPTIONS request") {
```
Comment (ericmeyer) [inline Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift]: This should be split into multiple tests. Remember, always ask yourself, can my test be testing less?

In this case, there should be three tests. One test per assertion.

File: Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import Quick
+import Nimble
+import Swerver
+import Foundation
+
+class NotAllowedActionSpec: QuickSpec {
+    
+    override func spec() {
+        describe("A Not Allowed action") {
+            let listOfAllowedMethods = ["GET", "POST"]
```
Comment (ericmeyer) [inline Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift]: I would avoid creating objects outside the scope of a test. I believe I made this comment in a different PR, but it was probably at the same time as you were already working on this new code.

File: Tests/SwerverTests/RouterSpec/RouterSpec.swift [hunk 9]
```diff
@@ -69,6 +70,18 @@ class RouterSpec: QuickSpec {
                     expect(router.process(request: anOptionsRequest)).to(equal(expectedResponse))
                 }
             }
+            
+            context("when receive an unknown request to an existing resource") {
```
Comment (ericmeyer) [inline Tests/SwerverTests/RouterSpec/RouterSpec.swift:106]: This test is testing too much for a unit test. (Although it is a reasonable function/higher-level test).

What is the failure message when they are not equal? What are some ways this test could be split?

**Review by ericmeyer — CHANGES_REQUESTED** [review]

File: Sources/SwerverCore/Action/HeadAction.swift (hunk 1 above)
Comment (mrfishball) [reply Sources/SwerverCore/Action/HeadAction.swift:4]: fixed

File: Sources/SwerverCore/Action/OptionsAction.swift (hunk 2 above)
Comment (mrfishball) [reply Sources/SwerverCore/Action/OptionsAction.swift]: will look into that

File: Sources/SwerverCore/Router/Route.swift (hunk 3 above)
Comment (mrfishball) [reply Sources/SwerverCore/Router/Route.swift:26]: 😄

File: Sources/SwerverCore/Router/Router.swift (hunk 4 above)
Comment (mrfishball) [reply Sources/SwerverCore/Router/Router.swift:28]: 😺

File: Sources/SwerverCore/Router/Routes.swift (hunk 5 above)
Comment (mrfishball) [reply Sources/SwerverCore/Router/Routes.swift:5]: 😺

File: Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift (hunk 6 above)
Comment (mrfishball) [reply Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift]: will look into that

File: Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift (hunk 7 above)
Comment (mrfishball) [reply Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift]: will look into that

File: Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift (hunk 8 above)
Comment (mrfishball) [reply Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift]: will look into other ways to test

File: Tests/SwerverTests/RouterSpec/RouterSpec.swift (hunk 9 above)
Comment (mrfishball) [reply Tests/SwerverTests/RouterSpec/RouterSpec.swift:106]: will look at splitting up this test and make it more granular

File: Sources/SwerverCore/Action/OptionsAction.swift (hunk 2 above)
Comment (mrfishball) [reply Sources/SwerverCore/Action/OptionsAction.swift]: implemented

File: Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift (hunk 6 above)
Comment (mrfishball) [reply Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift]: response.get implemented. Question: currently, the formatter class is able to loop through the headers dictionary and extract each header, however since the response.get(header: ...) func is now implemented, should I change the access identifier from public to private?

File: Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift (hunk 8 above)
Comment (mrfishball) [reply Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift]: fixed

File: Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift (hunk 7 above)
Comment (mrfishball) [reply Tests/SwerverTests/ActionSpec/NotAllowedActionSpec.swift]: this unit test has been broken down

File: Tests/SwerverTests/RouterSpec/RouterSpec.swift (hunk 9 above)
Comment (mrfishball) [reply Tests/SwerverTests/RouterSpec/RouterSpec.swift:106]: will look into that

Comment (mrfishball) [reply Tests/SwerverTests/RouterSpec/RouterSpec.swift:106]: all unit tests are further broken down

File: Sources/SwerverCore/Response/ResponseBuilder.swift [hunk 10]
```diff
@@ -5,7 +5,9 @@ public class ResponseBuilder {
     private(set) var body: String = String()
     private(set) var headers: [ResponseHeader:String] = [:]
 
-    public init() {}
+    public init() {
+        headers[.contentLength] = String(0)
```
Comment (ericmeyer) [inline Sources/SwerverCore/Response/ResponseBuilder.swift:9]: This should likely be called during `build`. That way you can incorporate the body if/when it is present.

File: Sources/SwerverCore/Router/Route.swift [hunk 11]
```diff
@@ -15,7 +15,7 @@ public class Route {
         for (method, _) in actions {
             listOfMethods.append(method.rawValue)
         }
-        return listOfMethods
+        return listOfMethods.sorted(by: <)
```
Comment (ericmeyer) [inline Sources/SwerverCore/Router/Route.swift:18]: It looks like you are sorting both here and in the options action. Are both `sorted` calls needed?

**Review by ericmeyer — APPROVED** [review]

File: Sources/SwerverCore/Action/OptionsAction.swift [hunk 12]
```diff
@@ -15,15 +15,17 @@ public class OptionsAction: HttpAction {
...
             .build()
     }
     
-    private func getAllowedMethods() -> String {
-        var allowedMethodList: [String] = []
-        for (method, _) in routes.fetchRoute(url: route) {
-            allowedMethodList.append(method.rawValue)
+    private func allowedMethods() -> String {
+        var listOfMethods = routes.allowedMethods(url: route)
+        if route.absoluteString == Resource.test_option_two.rawValue {
```
Comment (ericmeyer) [inline Sources/SwerverCore/Action/OptionsAction.swift:24]: For future reference, your last commit should probably have been its own pull request. It seems unrelated to this PR.

Comment (mrfishball) [reply Sources/SwerverCore/Action/OptionsAction.swift:24]: 👌

File: Sources/SwerverCore/Router/Route.swift (hunk 11 above)
Comment (mrfishball) [reply Sources/SwerverCore/Router/Route.swift:18]: I have since removed the sorting from options action and not allowed action

File: Sources/SwerverCore/Response/ResponseBuilder.swift (hunk 10 above)
Comment (mrfishball) [reply Sources/SwerverCore/Response/ResponseBuilder.swift:9]: I think setting it to "0" in the beginning makes sense because at any point .withBody is called, it'll set the content length and if I put it in .build() it'll reset the content length unless I use a conditional to check.

**Review by cmvandrevala — APPROVED** [review]
11 commits is a lot. Is there a place to rebase here, or is this purposefully split up into so many commits?
